File: june_challange/src/InvertBinaryTree.java
import java.util.Stack;

/**
 * https://leetcode.com/explore/challenge/card/june-leetcoding-challenge/539/week-1-june-1st-june-7th/3347/
 */
public class InvertBinaryTree {

    public TreeNode invertTree(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();

        if (root != null)
            stack.push(root);

        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            TreeNode temp = node.left;

            node.left = node.right;
            node.right = temp;

            if (node.left != null)
                stack.push(node.left);
            if (node.right != null)
                stack.push(node.right);
        }
        return root;
    }

    public TreeNode invertTree_recursion(TreeNode root) {
        if (root != null)
            exchange(root);
        return root;
    }

    void exchange(TreeNode node) {
        TreeNode temp = node.left;

        node.left = node.right;
        node.right = temp;

        if (node.left != null)
            exchange(node.left);
        if (node.right != null)
            exchange(node.right);
    }
    
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
           this.val = val;
           this.left = left;
           this.right = right;
        }
   }
}

File: june_challange/InvertBinaryTree.py
# https://leetcode.com/explore/challenge/card/june-leetcoding-challenge/539/week-1-june-1st-june-7th/3347/
class Solution:
    class TreeNode:
        def __init__(self, val=0, left=None, right=None):
            self.val = val
            self.left = left
            self.right = right

    def invertTree(self, root: TreeNode) -> TreeNode:

        stack = []
        if root is not None:
            stack.append(root)

        while (stack):
            node = stack.pop()
            temp = node.left
            node.left = node.right
            node.right = temp

            if node.left is not None:
                stack.append(node.left)

            if node.right is not None:
                stack.append(node.right)

        return root



    def invertTreeRec(self, root: TreeNode) -> TreeNode:
        if root is None:
            return root

        temp = root.left
        root.left = root.right
        root.right = temp

        self.invert(root.left)
        self.invert(root.right)
        return root
